### PR TITLE
KAFKA-15129;[6/N] Remove metrics in ControllerStats when broker shutdown

### PR DESCRIPTION
**Pull request by hudeqi**
This pr is used to remove the metrics in ControllerStats when broker shutdown.
This pr has passed the corresponding unit test, and it is part of [KAFKA-15129](https://issues.apache.org/jira/browse/KAFKA-15129).

### CODE REVIEW

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 1]
```diff
@@ -2788,6 +2796,20 @@ private[controller] class ControllerStats {
   def removeMetric(name: String): Unit = {
     metricsGroup.removeMetric(name)
   }
+
+  def removeMetrics(): Unit = {
+    MeterMetricNames.foreach(metricsGroup.removeMetric(_))
+    timerMetricNames.asScala.foreach(metricsGroup.removeMetric(_))
```
Comment (divijvaidya) [inline core/src/main/scala/kafka/controller/KafkaController.scala:2818]: remove entry from timerMetricNames as well else it keeps growing

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 2]
```diff
@@ -537,6 +538,7 @@ class KafkaController(val config: KafkaConfig,
 
   private def removeMetrics(): Unit = {
     KafkaController.MetricNames.foreach(metricsGroup.removeMetric)
+    controllerContext.stats.removeMetrics()
```
Comment (divijvaidya) [inline core/src/main/scala/kafka/controller/KafkaController.scala]: we should be registering these metrics in initializeControllerContext and removing them in resetControllerContext. The responsibility of correctly resetting and initializing controllerContext is that of KafkaController.

Comment (hudeqi) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: My thought is: this PR is to remove the ControllerStats metric when the broker shutdown. Before that, every broker in the cluster has registered all the metrics of ControllerStats. If following your suggestion, then only the real controller broker will register the metrics of ControllerStats, while other brokers do not. I don’t know this Is a logical change of a piece necessary? @divijvaidya

Comment (hudeqi) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: I first put the registration and removal of the metric of `ControllerStats` into `initializeControllerContext` and `resetContext`.
In addition, I also called `resetContext` in all the places where the `ControllerContext` was instantiated in the unit tests, so as to remove the `ControllerStats` metric.

**Pull request closed without merging**